src/vse_sync_pp/analyzers/pmc.py

```diff
@@ -0,0 +1,112 @@
+    STATE_HOLDOVER_IN_SPEC: [STATE_LOCKED, 
+                             STATE_HOLDOVER_IN_SPEC, 
+                             STATE_HOLDOVER_OUT_OF_SPEC1, 
+                             STATE_HOLDOVER_OUT_OF_SPEC2, 
+                             STATE_HOLDOVER_OUT_OF_SPEC3],
+    STATE_HOLDOVER_OUT_OF_SPEC1: [STATE_LOCKED, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC1, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC2, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC3],
+    STATE_HOLDOVER_OUT_OF_SPEC2: [STATE_LOCKED, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC1, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC2, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC3],
+    STATE_HOLDOVER_OUT_OF_SPEC3: [STATE_LOCKED, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC1, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC2, 
+                                  STATE_HOLDOVER_OUT_OF_SPEC3],
+}
+
+def is_transition_allowed(current_state, new_state):
+   return new_state in STATE_TRANSITION[current_state]
+
+class ClockStateAnalyzer(Analyzer):
+    """Analyze clock state
+    """
+    id_ = 'phc/gm-settings'
+    parser = id_
+    def __init__(self, config):
+        super().__init__(config)
+        # minimum test duration for a valid test
+        self._duration = config.parameter('min-test-duration/s')
+        self.state = None
+        self.transition_count = 0
+        self.clock_class_count = {
+            STATE_FREERUN: 0,
+            STATE_LOCKED: 0,
+            STATE_HOLDOVER_IN_SPEC: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC1: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC2: 0,
+            STATE_HOLDOVER_OUT_OF_SPEC3: 0,
+        }
+    def prepare(self, rows):
+        idx = 0
+        try:
+            tstart = rows[0].timestamp
+        except IndexError:
+            pass
+        else:
+            while idx < len(rows):
+                if tstart <= rows[idx].timestamp:
+                    break
+                idx += 1
+        return super().prepare(rows[idx:])
+    def test(self, data):
+        if len(data) == 0:
+            return (False, "no data")
+        
+        for index, row in data.iterrows():
+            clock_class = row['clock_class']
+ 
+            if (self.state is None) and (clock_class in STATE_TRANSITION):
+                self.state = clock_class
+            else:
+                if clock_class != self.state:
+                    self.transition_count += 1
+                
+                if clock_class in STATE_TRANSITION:
+                    transition_allowed = is_transition_allowed(self.state, clock_class)
+                else:
+                    return (False, f"wrong clock class {clock_class}")
+                
+                if transition_allowed:
+                    self.state = clock_class
+                    self.clock_class_count[clock_class] += 1
```

Comment:
Should we really be counting the the number of times we see a state or the number of transitions into the state? If nothing changes we will be just counting the number of samples? 

@jnunyez What do you think here?

Reply:
That was Jose's requirement 😄

Reply:
I just had a chat with Jose. I convinced him that counting the transition from `X->Y` is better. Where `X` is `self.state` and `Y` is `clock_class`. We will still count when `X` and `Y` are the same state but we will also have the info when they are not. I would suggest something like a `defaultdict` here as it will prevent having to initialize every transition.

Reply:
I think an improvement could be to have both: number of time we see state and number of times we see a state transitions.  
That will get us closer to the information that a plot can provide.

Reply:
We already have something like this. This is the current output:
```json
{"result": true, "reason": null, "analysis": {"duration": 1999.0688657, "clock_class_count": {"FREERUN": 0, "LOCKED": 500, "HOLDOVER_IN_SPEC": 1008, "HOLDOVER_OUT_SPEC": 488}, "total_transitions": 2}}
```

Reply:
Also if you do something like 
```python
STATE_NAMES = {
   STATE_FREERUN: "FREERUN",
   STATE_LOCKED: "LOCKED",
   STATE_HOLDOVER_IN_SPEC: "HOLDOVER_IN_SPEC"
   STATE_HOLDOVER_OUT_OF_SPEC1: "HOLDOVER_OUT_SPEC"
   STATE_HOLDOVER_OUT_OF_SPEC2: "HOLDOVER_OUT_SPEC"
   STATE_HOLDOVER_OUT_OF_SPEC3: "HOLDOVER_OUT_SPEC"
}
```
You can have can something like 
```python
self.transition_count = defaultdict(int)
self.transition_count[f"{STATE_NAMES[self.state]}->{STATE_NAMES[clock_class]}"] += 1
```

Reply:
Suggesting using comprehension instead for readability
```python
self.clock_class_count = {STATE_NAMES[k]:v for (k,v) in self.clock_class_count.items()}
```

The manipulation for the out-of-spec is only temporary until we have an answer about which value we should expect

Reply:
> We already have something like this. This is the current output:
> 
> ```json
> {"result": true, "reason": null, "analysis": {"duration": 1999.0688657, "clock_class_count": {"FREERUN": 0, "LOCKED": 500, "HOLDOVER_IN_SPEC": 1008, "HOLDOVER_OUT_SPEC": 488}, "total_transitions": 2}}
> ```

That only shows the end state, the starting state is useful. `Holdover in spec` to `locked` is different from `holdover out of spec` to `locked`

Reply:
@jnunyez Do you see a value in adding this? it will make the output pretty big. 
Something like this, but for every state:
```json
"clock_class_count": {"FREERUN": { "count": 0, "transitions": {"FREERUN": 10, "LOCKED": 1, "HOLDOVER_IN_SPEC": 0, "HOLDOVER_OUT_SPEC": 0}}}
```

Reply:
Looks good to me. That jointly with the plot will provide all the info to interpret that data.

Reply:
I pushed a proposal for this, let me know what you think. Maybe it's a bit dirty.
```json
{"result": true, "reason": null, "analysis": {"duration": 1999.0688657, "clock_class_count": {"FREERUN": {"count": 0, "transitions": {"FREERUN": 0, "LOCKED": 0, "HOLDOVER_IN_SPEC": 0, "HOLDOVER_OUT_SPEC1": 0, "HOLDOVER_OUT_SPEC2": 0, "HOLDOVER_OUT_SPEC3": 0}}, "LOCKED": {"count": 500, "transitions": {"FREERUN": 0, "LOCKED": 500, "HOLDOVER_IN_SPEC": 1, "HOLDOVER_OUT_SPEC1": 0, "HOLDOVER_OUT_SPEC2": 0, "HOLDOVER_OUT_SPEC3": 0}}, "HOLDOVER_IN_SPEC": {"count": 1008, "transitions": {"FREERUN": 0, "LOCKED": 0, "HOLDOVER_IN_SPEC": 1007, "HOLDOVER_OUT_SPEC1": 1, "HOLDOVER_OUT_SPEC2": 0, "HOLDOVER_OUT_SPEC3": 0}}, "HOLDOVER_OUT_SPEC1": {"count": 488, "transitions": {"FREERUN": 0, "LOCKED": 0, "HOLDOVER_IN_SPEC": 0, "HOLDOVER_OUT_SPEC1": 487, "HOLDOVER_OUT_SPEC2": 0, "HOLDOVER_OUT_SPEC3": 0}}, "HOLDOVER_OUT_SPEC2": {"count": 0, "transitions": {"FREERUN": 0, "LOCKED": 0, "HOLDOVER_IN_SPEC": 0, "HOLDOVER_OUT_SPEC1": 0, "HOLDOVER_OUT_SPEC2": 0, "HOLDOVER_OUT_SPEC3": 0}}, "HOLDOVER_OUT_SPEC3": {"count": 0, "transitions": {"FREERUN": 0, "LOCKED": 0, "HOLDOVER_IN_SPEC": 0, "HOLDOVER_OUT_SPEC1": 0, "HOLDOVER_OUT_SPEC2": 0, "HOLDOVER_OUT_SPEC3": 0}}}, "total_transitions": 2}}
```

Reply:
I like it. A few comments:

- Just add the information for those states where count is bigger than zero. If for a state `count` is Zero don't add that information for the state since we already know transitions from that state are going to be zero. 

- Doubt: In the example above if `count` for "HOLDOVER_IN_SPEC" is 1008 why transitions to HOLDOVER_IN_SPEC is 1007?

Reply:
> * Just add the information for those states where count is bigger than zero. If for a state `count` is Zero don't add that information for the state since we already know transitions from that state are going to be zero.

I think we should keep the output the same for every run, even if it with zero occurrences. @nocturnalastro wdyt?
> * Doubt: In the example above if `count` for "HOLDOVER_IN_SPEC" is 1008 why transitions to HOLDOVER_IN_SPEC is 1007?

That's because there are 1007 transitions from `HOLDOVER_IN_SPEC` to `HOLDOVER_IN_SPEC`. You can see 1 more transition from `LOCKED`

Reply:
Aha got it. Thanks! :-)

Reply:
> I think we should keep the output the same for every run, even if it with zero occurrences. @nocturnalastro wdyt?

If that simplifies the generation of report using tables then it makes sense to me.

Reply:
Yeah, just use a defaultdict then you only have the ones you've accessed :)